lib/recursive-methods.rb

```diff
@@ -1,49 +1,130 @@
+    end
+    if n == 0
+      return 1;
+    end 
+    return n * factorial(n-1);
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: o(n)
+# Space complexity: o(n) 
+def reverse(s, current = s.length-1, reversed_string="")
+  if current >= 0
+    reversed_string << s[current] 
+    current -= 1
+    reverse(s, current, reversed_string)
+  else
+    return reversed_string
+  end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: o(n)
+# Space complexity: o(n) 
+def reverse_inplace(s, start=0)
+  if start <= ((s.length-1)/2)  
+    s[start], s[s.length - 1 - start] = s[s.length - 1 - start], s[start]
+    start+=1
+    reverse_inplace(s, start)
+  end 
+  return s
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n) 
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 0
+    return 0
+  else
+    return 2 + bunny(n-1)
+  end   
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n) 
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length%2 != 0 
+    return false
+  end
+  if s.length == 0 
+    return true
+  end
+
+  start = 0
+  ending = s.length - 1
+  while start < ending
+    if s[start] != "(" || s[ending] != ")" 
+      return false;
+    end
+    start += 1
+    ending =- 1
+    return nested((s[start...ending]));
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: o(n)
+# Space complexity: o(n)
+def search(array, value, start=0)
+  if start < array.length  
+    if value == array[start] 
+        return true
+    end
+    start+=1
+    return search(array, value, start)
+  end
+  return false
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n), each character has a stack
 def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length <= 1
+      return true
+  end 
+  start = 0
+  ending = s.length - 1
+  while start < ending
+    if s[start] != s[ending]
+      return false
+    end
+    start += 1
+    ending =- 1
+    return is_palindrome(s[start...ending]);
+  end
+end
+
+# Time complexity: o(log(n)) 
+# Space complexity: o(n + m), n and m have their own "stacks"
+def digit_match(n, m, count=0)
+
+  if n == 0 && m == 0 
+    return 1;
+  end
+
+  if n%10 == m%10
+    count+=1;
+  end 
+
+  if n/10 == 0 || m/10 == 0
+    return count
+  end 
+
+  return digit_match(n/10, m/10, count)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def digit_match(n, m)
-    raise NotImplementedError, "Method not implemented"
-end
\ No newline at end of file
+
+#time: o(n)
+#space: o(n)
+
+def fib(n)
```

Comment:
Notice that you have 2 recursive calls each time.  This is actually O(2^n) in time complexity.